### PR TITLE
Make association queries to preparable

### PR DESCRIPTION
**Pull request by kamipo**
``` ruby
  # Before
  PriceEstimate.where(estimate_of: treasure).to_a
  # => SELECT "price_estimates".* FROM "price_estimates" WHERE "price_estimates"."estimate_of_type" = 'Treasure' AND "price_estimates"."estimate_of_id" = 1

  # After
  PriceEstimate.where(estimate_of: treasure).to_a
  # => SELECT "price_estimates".* FROM "price_estimates" WHERE "price_estimates"."estimate_of_type" = ? AND "price_estimates"."estimate_of_id" = ?  [["estimate_of_type", "Treasure"], ["estimate_of_id", 1]]
```

### CODE REVIEW

File: activerecord/lib/active_record/relation/predicate_builder.rb [hunk 1]
```diff
@@ -107,25 +107,33 @@ def create_binds_for_hash(attributes)
...
+        end
+
+        [parts, binds]
+      end
+
+      def build_for_association_query(column_name, value)
+        parts, binds = [], []
+        AssociationQueryValue.queries_for(table, column_name, value).each do |query|
+          prts, bvs = build_from_hash(query)
+          parts << prts
```
Comment (esparta) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: can this use `parts.concat` for the sake of use the consistency?

Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: Cannot use `parts.concat`. `binds` should be flatten, but `parts` should be kept `type` and `id` pair.

```ruby
parts = [[:prts1_type, :prts2_id], [:prts2_type, :prts2_id]]
binds = [:bvs1_type, :bvs1_id, :bvs2_type, :bvs2_id]
```

**Comment by kamipo** [comment]
Fixes #26541.

**Comment by kamipo** [comment]
This has been almost resolved in #28715 (remaining parts are #26098 and #28539).

**Pull request closed without merging**